### PR TITLE
PIA-658: extend sign in tests coverage

### PR DESCRIPTION
**Pull request by kp-ruth-ondrada**
## Summary

We are adding 2 tests cases for sign in 

1. When user has no password provided
2. When user no username provided

The expected results is to provide an error message to the user informing him that he needs both a valid username and password.

We've done some refactoring of `enterCredentials` to facilitate login maintainability.
We've created a constant for `defaultTimeout` period.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -238,6 +238,8 @@ android {
             debuggable true
             buildConfigField 'String', 'PIA_VALID_USERNAME', "\"" + System.getenv("PIA_VALID_USERNAME") + "\""
             buildConfigField 'String', 'PIA_VALID_PASSWORD', "\"" + System.getenv("PIA_VALID_PASSWORD") + "\""
+            buildConfigField 'String', 'PIA_INVALID_USERNAME', "\"${System.getenv("PIA_INVALID_USERNAME")}\""
```
Comment (kp-aldo-pedromingo) [inline app/build.gradle:241]: In order to standardize. Can we use the same format as the other fields?

```
buildConfigField 'String', 'PIA_INVALID_USERNAME', "\"" + System.getenv("PIA_INVALID_USERNAME") + "\""
buildConfigField 'String', 'PIA_INVALID_PASSWORD', "\"" + System.getenv("PIA_INVALID_PASSWORD") + "\""
```

File: app/build.gradle [hunk 2]
```diff
@@ -235,6 +236,11 @@ android {
         }
         debug {
             debuggable true
+            buildConfigField 'String', 'PIA_VALID_USERNAME', "\"" + System.getenv("PIA_VALID_USERNAME") + "\""
+            buildConfigField 'String', 'PIA_VALID_PASSWORD', "\"" + System.getenv("PIA_VALID_PASSWORD") + "\""
+            buildConfigField 'String', 'PIA_INVALID_USERNAME', "\"${System.getenv("PIA_INVALID_USERNAME")}\""
+            buildConfigField 'String', 'PIA_INVALID_PASSWORD', "\"${System.getenv("PIA_INVALID_PASSWORD")}\""
```
Comment (kp-juan-docal) [inline app/build.gradle:242]: Any particular reason why we are doing on some:

`${System.getenv("PIA_INVALID_PASSWORD")}` 

and on others:

`+ System.getenv("PIA_VALID_PASSWORD") +` 

If there is no particular preference, I'd say let's commit to one for consistency purposes. What do you think?

**Review by kp-juan-docal — DISMISSED** [review]
nit. LGTM!

File: app/build.gradle (hunk 1 above)
Comment (kp-aldo-pedromingo) [reply app/build.gradle:241]: As an alternative. For invalid data, we don't necessarily need to add them as a secret.
The reason why we add `VALID` data here is to prevent open source users to get usable credentials. For invalid credentials we can just use garbage strings which don't need to be protected from anybody :D

File: app/build.gradle (hunk 2 above)
Comment (kp-ruth-ondrada) [reply app/build.gradle:242]: hey @kp-juan-docal , 

thank's for looking into it! 
For those that are using `+ System.getenv("PIA_VALID_PASSWORD") +` @kp-andrei-madescu  already has this change in his other branch.

TLDR;
We will use `${System.getenv("PIA_INVALID_PASSWORD")}` and some changes are in Andrei's 🙏 

hope that's okays!

File: app/src/androidTest/java/tests/SignInTests.kt [hunk 3]
```diff
@@ -39,4 +39,24 @@ class SignInTests : BaseUiAutomatorClass() {
         stepObjects.clickOnLoginButton()
         assert(SignInPageObjects().noUsernameOrPasswordError.exists())
     }
+
+    @Test
+    fun errorMessageIfNoPasswordProvided() {
```
Comment (kp-aldo-pedromingo) [inline app/src/androidTest/java/tests/SignInTests.kt:45]: Cool stuff!

File: app/src/androidTest/java/tests/SignInTests.kt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        stepObjects.allowVpnProfileCreation()
+        assert(MainScreenPageObjects().connectButton.exists())
+    }
+
+    @Test
+    fun incorrectCredentialsReturnToSplashScreen() {
+        stepObjects.allowNotifications()
+        stepObjects.reachSignInScreen()
+        stepObjects.enterUsername(BuildConfig.PIA_INVALID_USERNAME)
+        stepObjects.enterPassword(BuildConfig.PIA_INVALID_PASSWORD)
```
Comment (kp-gabriel-sasu) [inline app/src/androidTest/java/tests/SignInTests.kt]: I like that you've built dedicated step-objects for various action groups I would suggest a change here however. I would recommend having this function parameterised with a `username` and `password` and having the default value set to your env variables. This is helpful if in the future you want to have a particular case covered i.e: Special case characters in the username and passwords fields without having to change the env variables or create new ones

File: app/src/androidTest/java/tests/SignInTests.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        stepObjects.reachSignInScreen()
+        stepObjects.clickOnLoginButton()
+        assert(SignInPageObjects().noUsernameOrPasswordError.exists())
+    }
+
+    @Test
+    fun errorMessageIfNoPasswordProvided() {
+        stepObjects.allowNotifications()
+        stepObjects.reachSignInScreen()
+        stepObjects.enterUsername(BuildConfig.PIA_VALID_USERNAME)
```
Comment (kp-gabriel-sasu) [inline app/src/androidTest/java/tests/SignInTests.kt]: Similar with the previous comment on this file, for future proofing this, a parameter with default value is a lot more useful in the long run

**Review by kp-aldo-pedromingo — COMMENTED** [review]
The commit message is good but could be more descriptive.
It is considered good practice to use the commit message as support documentation.
It doesn't need to be lengthy. You can take this as commit from our desktop bros as inspiration: https://github.com/xvpn/pia_desktop/pull/1224/commits/42c0bee1f083d342b3de82a0798d23683d9bbfae

File: app/build.gradle (hunk 2 above)
Comment (kp-juan-docal) [reply app/build.gradle:242]: absolutely! thanks! 🙇

File: app/src/androidTest/java/screens/steps/SignInStepObjects.kt [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    fun enterUsername(username : String?) {
+        inputTextInField(signInPageObjects.usernameField, username)
+    }
+
+    fun enterPassword(password : String?) {
+        inputTextInField(signInPageObjects.passwordField, password)
+    }
+
+    fun clickOnLoginButton() {
+        signInPageObjects.loginButton.clickAndWaitForNewWindow(5000)
```
Comment (kp-gabriel-sasu) [inline app/src/androidTest/java/screens/steps/SignInStepObjects.kt]: Since both line 28 and line 16 employ the same time based parameter for click and wait action, I would recommend having the value set as a constant in the file or in the base class rather than hardcoded value. This improves maintenance long term and simplifies refactoring if needed.

File: app/src/androidTest/java/screens/objects/SignInPageObjects.kt [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+import com.privateinternetaccess.android.helpers.LocatorHelper
+
+class SignInPageObjects {
+
+    val notificationPrompt = LocatorHelper.findByResourceId("com.android.permissioncontroller:id/permission_message")
+    val allowNotifications = LocatorHelper.findByResourceId("com.android.permissioncontroller:id/permission_allow_button")
+    val denyNotifications = LocatorHelper.findByResourceId("com.android.permissioncontroller:id/permission_deny_button")
+    val reachLoginScreenButton = LocatorHelper.findByResourceId("com.privateinternetaccess.android:id/login")
+    val usernameField = LocatorHelper.findByResourceId("com.privateinternetaccess.android:id/piaxEditText")
+    val passwordField = LocatorHelper.findByResourceId("com.privateinternetaccess.android:id/piaxEditText", instance = 1)
```
Comment (kp-gabriel-sasu) [inline app/src/androidTest/java/screens/objects/SignInPageObjects.kt:12]: A bit out of scope of this PR however I would suggest considering the addition a dedicated locator/resourceId for the passwordField since indexes/instances can cause instability later down the line if the element order changes or if new elements are added

File: app/build.gradle (hunk 1 above)
Comment (kp-ruth-ondrada) [reply app/build.gradle:241]: Fixed this one for them to be all same format

For invalid ones, we want to have consistency so we will keep them as they are for the time being

File: app/src/androidTest/java/screens/steps/SignInStepObjects.kt (hunk 6 above)
Comment (kp-ruth-ondrada) [reply app/src/androidTest/java/screens/steps/SignInStepObjects.kt]: We declared a companion object with const value with 5000L.
We import it in the `SignInStepsObjects` to improve readability and maintenance.

File: app/src/androidTest/java/screens/objects/SignInPageObjects.kt (hunk 7 above)
Comment (kp-ruth-ondrada) [reply app/src/androidTest/java/screens/objects/SignInPageObjects.kt:12]: We will take note of this as right now its out of scope but indeed it's good to have to avoid flakiness.
It will be handled in another pr

File: app/src/androidTest/java/tests/SignInTests.kt (hunk 4 above)
Comment (kp-ruth-ondrada) [reply app/src/androidTest/java/tests/SignInTests.kt]: We've refactored the enterCredentials method to take as default values the valid username and password with the possibility to override them.

File: app/src/androidTest/java/tests/SignInTests.kt (hunk 5 above)
Comment (kp-ruth-ondrada) [reply app/src/androidTest/java/tests/SignInTests.kt]: same as above, we've refactored the enterCredentials.

**Comment by kp-ruth-ondrada** [comment]
> The commit message is good but could be more descriptive. It is considered good practice to use the commit message as support documentation. It doesn't need to be lengthy. You can take this as commit from our desktop bros as inspiration: [xvpn/pia_desktop@42c0bee](https://github.com/xvpn/pia_desktop/commit/42c0bee1f083d342b3de82a0798d23683d9bbfae)

I've updated the summary to be more descriptive.
@kp-aldo-pedromingo  @kp-gabriel-sasu , can you help to review again. Thank you 🙏

File: app/src/androidTest/java/screens/steps/SignInStepObjects.kt [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+    fun allowNotifications() {
+        clickIfExists(signInPageObjects.notificationPrompt, signInPageObjects.allowNotifications)
+    }
+
+    fun reachSignInScreen() {
+        signInPageObjects.reachLoginScreenButton.clickAndWaitForNewWindow(defaultTimeOut)
+    }
+
+    fun enterCredentials(username : String = BuildConfig.PIA_VALID_USERNAME, password : String = BuildConfig.PIA_VALID_PASSWORD) {
```
Comment (kp-gabriel-sasu) [inline app/src/androidTest/java/screens/steps/SignInStepObjects.kt]:
```suggestion
    fun enterCredentials(
        username : String = BuildConfig.PIA_VALID_USERNAME, 
        password : String = BuildConfig.PIA_VALID_PASSWORD
    ) {
```

File: app/src/androidTest/java/tests/SignInTests.kt [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        stepObjects.clickOnLoginButton()
+        stepObjects.allowVpnProfileCreation()
+        assert(MainScreenPageObjects().connectButton.exists())
+    }
+
+    @Test
+    fun incorrectCredentialsReturnToSplashScreen() {
+        stepObjects.allowNotifications()
+        stepObjects.reachSignInScreen()
+        stepObjects.enterCredentials(BuildConfig.PIA_INVALID_USERNAME, BuildConfig.PIA_INVALID_PASSWORD)
```
Comment (kp-gabriel-sasu) [inline app/src/androidTest/java/tests/SignInTests.kt]:
```suggestion
        stepObjects.enterCredentials()
```

Since your function now has default values, it is no longer needed to pass in the env vars unless you want to be very explicit. It's needed to pass in values for `enterCredentials` call only if you want to pass in values that are different than the default ones set

**Review by kp-andrei-madescu — APPROVED** [review]
Looks good and the changes made will improve the maintainability in the long term, which is always a good thing. 
Great work and congrats on the first PR! 🚀

**Comment by kp-gabriel-sasu** [comment]
Woo! Congrats Ruth 🚀 👏 🎉  A wonderful and lovely PR, very nicely done! To many more ahead 🥳
